x-pack/libbeat/reader/etw/session.go

```diff
@@ -0,0 +1,265 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build windows
+
+package etw
+
+import (
+	"errors"
+	"fmt"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+// For testing purposes we create a variable to store the function to call
+// When running tests, these variables point to a mock function
+var (
+	GUIDFromProviderNameFunc = GUIDFromProviderName
+	SetSessionGUIDFunc       = setSessionGUID
+)
+
+type Session struct {
+	// Name is the identifier for the session.
+	// It is used to identify the session in logs and also for Windows processes.
+	Name string
+	// GUID is the provider GUID to configure the session.
+	GUID GUID
+	// Properties of the session that are initialized in newSessionProperties()
+	// See https://learn.microsoft.com/en-us/windows/win32/api/evntrace/ns-evntrace-event_trace_properties for more information
+	Properties *EventTraceProperties
+	// Handler of the event tracing session for which the provider is being configured.
+	// It is obtained from StartTrace when a new trace is started.
+	// This handler is needed to enable, query or stop the trace.
+	Handler uintptr
+	// Realtime is a flag to know if the consumer reads from a logfile or real-time session.
+	Realtime bool // Real-time flag
+	// NewSession is a flag to indicate whether a new session has been created or attached to an existing one.
+	NewSession bool
+	// TraceLevel sets the maximum level of events that we want the provider to write.
+	TraceLevel uint8
+	// MatchAnyKeyword is a 64-bit bitmask of keywords that determine the categories of events that we want the provider to write.
+	// The provider writes an event if the event's keyword bits match any of the bits set in this value
+	// or if the event has no keyword bits set, in addition to meeting the Level and MatchAllKeyword criteria.
+	MatchAnyKeyword uint64
+	// MatchAllKeyword is a 64-bit bitmask of keywords that restricts the events that we want the provider to write.
+	// The provider typically writes an event if the event's keyword bits match all of the bits set in this value
+	// or if the event has no keyword bits set, in addition to meeting the Level and MatchAnyKeyword criteria.
+	MatchAllKeyword uint64
+	// TraceHandler is the trace processing handle.
+	// It is used to control the trace that receives and processes events.
+	TraceHandler uint64
+	// Callback is the pointer to EventRecordCallback which receives and processes event trace events.
+	Callback uintptr
```

Comment:
I think it would be better for users of the API if instead of representing function pointers using `uintptr` it used strong typed Go function signatures (e.g. `func(*EventRecord)`).

Reply:
I don't find the way to apply this without getting things a bit more complicated. This callback is assigned calling `syscall.NewCallback` which returns an `uintptr` so to adapt that we would need a wrapper for the callback.